src/components/Main.tsx

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useCallback, useContext, useMemo, useState } from 'react';
+import { Todo } from '../types/Todo';
+import classNames from 'classnames';
+import { ErrorContext, LoadingContext, TodosContext } from './TodoContext';
+import { deleteTodos, editTodos } from '../api/todos';
+import { Status } from '../types/Status';
+import { getFilteredItems } from '../services/getFilteredItems';
+
+type Props = {
+  query: Status;
+};
+
+export const Main: React.FC<Props> = ({ query }) => {
+  const { list, setList } = useContext(TodosContext);
+  const { setErrorMessage } = useContext(ErrorContext);
+  const { isLoading } = useContext(LoadingContext);
+  const [editingTodo, setEditingTodo] = useState<Todo | null>(null);
+  const listToGo = useMemo(() => {
+    return getFilteredItems(list, query);
+  }, [list, query]);
+
+  const handleTodoDelete = useCallback(
+    (todoId: number) => {
+      deleteTodos(todoId).catch(() =>
+        setErrorMessage('Unable to delete a todo'),
+      );
+
+      setList(currentTodos => {
+        return currentTodos.filter(currentTodo => currentTodo.id !== todoId);
+      });
+    },
+    [setErrorMessage, setList],
+  );
+
+  const handleTodoUpdate = useCallback(
+    (updatedTodo: Todo) => {
+      if (!updatedTodo.title.trim()) {
+        handleTodoDelete(updatedTodo.id);
+        setErrorMessage('Unable to add a todo');
+
+        return;
+      }
+
+      editTodos(updatedTodo)
+        .then(() => {
+          setList(currentTodos =>
+            currentTodos.map(currentTodo =>
+              currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+            ),
+          );
+          setEditingTodo(null);
+        })
+        .catch(() => setErrorMessage('Unable to update a todo'));
+    },
+    [handleTodoDelete, setErrorMessage, setList],
+  );
+
+  const handleCheckbox = (todo: Todo) => {
+    const updatedList = list.map((item: Todo) => {
+      if (item.id === todo.id) {
+        return { ...item, completed: !item.completed };
+      }
+
+      return item;
+    });
+
+    setList(updatedList);
```

Comment:
When you need to change state depending on the previous one use functional setState